lib/postgrex/simple_connection.ex

```diff
@@ -0,0 +1,274 @@
+      case Keyword.get(opts(mod), :configure) do
+        {module, fun, args} -> apply(module, fun, [opts(mod) | args])
+        fun when is_function(fun, 1) -> fun.(opts(mod))
+        nil -> opts(mod)
+      end
+
+    case Protocol.connect(opts) do
+      {:ok, protocol} ->
+        state = %{state | protocol: protocol}
+
+        with {:noreply, state, _} <- maybe_handle(mod, :handle_connect, [mod_state], state) do
+          {:ok, state}
+        end
+
+      {:error, reason} ->
+        if state.auto_reconnect do
+          {:backoff, state.reconnect_backoff, state}
+        else
+          {:stop, reason, state}
+        end
+    end
+  end
+
+  @doc false
+  def handle_call(msg, from, %{state: {mod, mod_state}} = state) do
+    handle(mod, :handle_call, [msg, from, mod_state], from, state)
+  end
+
+  @doc false
+  def handle_info(:timeout, %{protocol: protocol} = state) do
+    case Protocol.ping(protocol) do
+      {:ok, protocol} ->
+        {:noreply, %{state | protocol: protocol}, state.idle_interval}
+
+      {error, reason, protocol} ->
+        reconnect_or_stop(error, reason, protocol, state)
+    end
+  end
+
+  def handle_info(msg, %{protocol: protocol, state: {mod, mod_state}} = state) do
+    opts = [notify: &mod.notify(&1, &2, mod_state)]
+
+    case Protocol.handle_info(msg, opts, protocol) do
+      {:ok, protocol} ->
+        {:noreply, %{state | protocol: protocol}, state.idle_interval}
+
+      {:unknown, protocol} ->
+        maybe_handle(mod, :handle_info, [msg, mod_state], %{state | protocol: protocol})
+
+      {error, reason, protocol} ->
+        reconnect_or_stop(error, reason, protocol, state)
+    end
+  end
+
+  def handle_info(msg, %{state: {mod, mod_state}} = state) do
+    maybe_handle(mod, :handle_info, [msg, mod_state], state)
+  end
+
+  defp maybe_handle(mod, fun, args, state) do
+    if function_exported?(mod, fun, length(args)) do
+      handle(mod, fun, args, nil, state)
+    else
+      {:noreply, state, state.idle_interval}
+    end
+  end
+
+  defp handle(mod, fun, args, from, state) do
+    case apply(mod, fun, args) do
+      {:noreply, mod_state} ->
+        {:noreply, %{state | state: {mod, mod_state}}, state.idle_interval}
+
+      {:query, query, mod_state} ->
+        opts = [notify: &mod.notify(&1, &2, mod_state)]
+
+        state = %{state | state: {mod, mod_state}}
+
+        with {:ok, result, protocol} <- Protocol.handle_simple(query, opts, state.protocol),
+             {:ok, protocol} <- Protocol.checkin(protocol) do
+          state = %{state | protocol: protocol}
+
+          handle(mod, :handle_result, [result, mod_state], from, state)
+        else
+          {:error, %Postgrex.Error{} = error, protocol} ->
+            handle(mod, :handle_result, [error, mod_state], from, %{state | protocol: protocol})
+
+          {:disconnect, reason, protocol} ->
+            from && reply(from, {__MODULE__, reason})
+
+            reconnect_or_stop(:disconnect, reason, protocol, state)
```

Comment:
```suggestion
          {:disconnect, reason, protocol} ->
            case reconnect_or_stop(:disconnect, reason, protocol, state) do
              {:stop, _, _} = stop ->
                stop

              {:connect, _, _} = connect ->
                from && reply(from, async_reply(mod_state, reason))
                connect
            end
        end
    end
  end

  defp async_reply(%{async_reply: reply} = mod_state, _reason), do: async_reply
  defp async_reply(_ = _mod_state, reason), do: {__MODULE__, reason}
```

@sorentwo I think you need something like this. It will fix your failing test and also it seems necessary for the notification module to function if the user tries to issue a listen command before the connection process handles the disconnect message from postgres.

Reply:
I think we should just remove the `from && ...` and document that, if auto_reconnect is true, you can fail and never get a reply unless you handle this case in `handle_disconnect`.

Reply:
Here is the approach I mentioned above implemented for replication: https://github.com/elixir-ecto/postgrex/commit/bf793fa20c8d7dfce1c5424ca5ea325d27e11e3d - I will push some docs soon.

Reply:
What do you think about changing the arity of `handle_disconnect` to 2 so it can receive whether or not the disconnect is a `:stop` or `:reconnect`? I think it might be necessary here so the Notifications module can reply with `:eventually` only on reconnect. But might be mistaken.

Reply:
@greg-rychlewski the module will always know if it is started with auto-reconnect or not, so it can figure it out if necessary. :)